### ISSUE TITLE
Notes

### ISSUE THREAD

**Issue by AntonC9018** [Artiom9/asdc_labs#1]
https://github.com/Artiom9/asdc_labs/blob/99df039b5205b515298652fb1807670b500689f5/lab2/lab2.cpp#L8
`const std::vector<T>&` -> `std::span<T>` или `std::span<const T>`

**Comment by AntonC9018** [comment]
Сами функции поиска нигде не юзаются?

**Comment by AntonC9018** [comment]
https://github.com/Artiom9/asdc_labs/blob/99df039b5205b515298652fb1807670b500689f5/lab2/lab2.cpp#L48
Тут конвертирование и сужение типа. Включи warning-и и не игнорируй их. Тут нужно использовать либо `size_t`, либо закастить вручную в `int`, чтобы было понятно, что теряется информация.

**Comment by AntonC9018** [comment]
https://github.com/Artiom9/asdc_labs/blob/99df039b5205b515298652fb1807670b500689f5/lab2/lab2.cpp#L106
Передай факт ошибки наружу тоже.

**Comment by AntonC9018** [comment]
serializator -> serializer

**Comment by AntonC9018** [comment]
https://github.com/Artiom9/asdc_labs/blob/99df039b5205b515298652fb1807670b500689f5/lab3/linked_list.cpp#L22
Добавь тут коммент, что идет в начало списка.

**Comment by AntonC9018** [comment]
https://github.com/Artiom9/asdc_labs/blob/99df039b5205b515298652fb1807670b500689f5/lab3/dlinked_list.cpp#L67-L70
Эти в принципе можно объединить, потому что первая и 3 проверка эквивалентные, как и вторая с 4

**Comment by AntonC9018** [comment]
https://github.com/biglevoozy/ASDC---Tsurtsurica/blob/1e5bbb62d2fccbcc8f5bad9bbee928d0e3d7453a/Library/HumanFncts.cpp#L54-L67
Зачем `std::string`? Ты же просто печатаешь это, зачем память выделять? `std::string_view`

**Comment by AntonC9018** [comment]
https://github.com/biglevoozy/ASDC---Tsurtsurica/blob/1e5bbb62d2fccbcc8f5bad9bbee928d0e3d7453a/Library/HumanFncts.cpp#L71
Удали откоменченный код.

**Comment by AntonC9018** [comment]
https://github.com/biglevoozy/ASDC---Tsurtsurica/blob/1e5bbb62d2fccbcc8f5bad9bbee928d0e3d7453a/Library/HumanFncts.cpp#L165
А остальное что? Просто игнорится? Если ты передашь `1dsffds0`, он это запарсит как `1`. Это действительно желаемый результат?

**Comment by AntonC9018** [comment]
https://github.com/biglevoozy/ASDC---Tsurtsurica/blob/1e5bbb62d2fccbcc8f5bad9bbee928d0e3d7453a/Library/ListFncts.cpp#L30
Только щас посмторел, что у тебя max_size для строки идет 999. Если так много символов позволяешь, копировать в стек такое будет дорого. Передавай референсом или указателем если можешь. Но вообще правильнее всего для такого огромного объекта было бы вернуть указатель, и туда вписать инфу, чтобы не копировать по возможности. Или сделай меньше строку. 999 слишком много имо.

**Comment by AntonC9018** [comment]
https://github.com/biglevoozy/ASDC---Tsurtsurica/blob/1e5bbb62d2fccbcc8f5bad9bbee928d0e3d7453a/Lab_7_Sorting_Algorithms/main.cpp#L43
span

**Comment by AntonC9018** [comment]
В доки с анализом вставь конкретно какие результаты получил. Скопируй из консоли сколько секунд занял алгоритм, сколько операций сравнения, т.п., для разных размеров массива. И из этого выведи сложность алгоритма.

**Comment by AntonC9018** [comment]
Сорри, не туда писал, щас удалю

**Comment by Artiom9** [comment]
> https://github.com/Artiom9/asdc_labs/blob/99df039b5205b515298652fb1807670b500689f5/lab3/dlinked_list.cpp#L67-L70
> 
> 
> Эти в принципе можно объединить, потому что первая и 3 проверка взаимопротивоположные, как и вторая с 4

Мне нужны 3 и 4 проверка для обновление внутренних переменных моего двусвязного списка, а 1 и 2 для управления связью между нодами. Я не понимаю, как они взаимосвязаны.

**Comment by Artiom9** [comment]
Остальное пофиксил. Спасибо за ваше время!

**Comment by AntonC9018** [comment]
> Насколько я понял, можно было создать свою структуру или использовать `std::optional`, а через него вроде нельзя возвращать код ошибки.
`std::optional` тут не очень пойдет, потому что у тебя логика такая, если я не ошибаюсь, что даже если были ошибки, возвращаются ряды без ошибок. Тут выбор или своя структура с булос или кодом ошибки, в зависимости от потребностей, или передавать указатель/референс на вектор в который писать результаты как параметр, или передавать указатель, куда вписать ошибку, параметром.

**Comment by Artiom9** [comment]
Ну да всё правильно, я выбрал концепцию, чтоб мы продолжали работу, даже если были ошибки. Я так понимаю, вы хотите, чтоб снаружи можно было узнать, если были какие-то ошибки при чтении файла?

**Comment by AntonC9018** [comment]
> Мне нужны 3 и 4 проверка для обновление внутренних переменных моего двусвязного списка, а 1 и 2 для управления связью между нодами. Я не понимаю, как они взаимосвязаны.

Ну типа если у нода нету `next`, то он является последним в списке. И наоборот, если нет `prev`, то он будет первым. Но не суть важно, тут для понятности можно оставить как есть в принципе, а можно дописать комментом или через ассерт второй проверки то, что я предложил. Делай как хочешь

**Comment by AntonC9018** [comment]
> Ну да всё правильно, я выбрал концепцию, чтоб мы продолжали работу, даже если были ошибки. Я так понимаю, вы хотите, чтоб снаружи можно было узнать, если были какие-то ошибки при чтении файла?

да

**Comment by Artiom9** [comment]
> > Мне нужны 3 и 4 проверка для обновление внутренних переменных моего двусвязного списка, а 1 и 2 для управления связью между нодами. Я не понимаю, как они взаимосвязаны.
> 
> Ну типа если у нода нету `next`, то он является последним в списке. И наоборот, если нет `prev`, то он будет первым. Но не суть важно, тут для понятности можно оставить как есть в принципе, а можно дописать комментом или через ассерт второй проверки то, что я предложил. Делай как хочешь

Нет, это я понял, но смотрите у меня есть скажем так две структуры, одна сама нода, в которые есть ссылки на следующую и предыдущую ноду, и внутри листа у меня есть два указателя на начало списка и конец, которые используется в `insertAfter` и `instertBefore`, когда передается `nullptr`. И получается мне за этими указателями тоже надо следить и обновлять их при удалении ноды

**Comment by AntonC9018** [comment]
Можешь предположить, что переданный параметром нод является нодом из списка. Иначе операция и сама по себе неверная -- пытаешься добавить в список после нола из другого списка, по сути. А если переданный нод это часть списка, то в случае если у него нету next он будет всегда равен tail, потому что у любого другого нода из списка next есть. И так же для начала списка. У тебя удаление? В целом, неважно какая операция. Всегда идет предположение, что нод из переданного списка.

**Comment by Artiom9** [comment]
Всё, я понял, надо добавить в 1 и 2 условия else и убрать проверки 3 и 4. Сейчас поправлю
